src/Spice86.Core/Emulator/OperatingSystem/DosMemoryManager.cs

```diff
@@ -162,10 +182,235 @@ public DosErrorCode TryModifyBlock(in ushort blockSegment, in ushort requestedSi
+                // This is what real DOS does. It always finds the largest free block that it can
+                // allocate. It's simple and relatively easy.
+                DosMemoryControlBlock largestFreeBlock = FindLargestFree();
+                if (largestFreeBlock.IsValid) {
+                    freeSizeInParagraphs = largestFreeBlock.Size;
+                }
+            } else {
+                // This behavior is unique to Spice86. Real DOS doesn't support loading a program at
+                // a specific address. It always allocates a new block for it. However to support
+                // loading a program at the location specified in the configuration at
+                // initialization time, we support it. It's handy for reverse engineering to ensure
+                // that the load address is always the same. Rather than finding the largest free
+                // block, we just allocate the size of the block at the given address as long as it
+                // is free.
+                DosMemoryControlBlock requestedBlock = GetDosMemoryControlBlockFromSegment(
+                    (ushort)(pspSegment - 1));
+                if (requestedBlock.IsValid && requestedBlock.IsFree) {
+                    freeSizeInParagraphs = requestedBlock.Size;
+                }
+            }
+
+            // It's possible that we didn't find a suitable free block, or that the block that we
+            // found isn't large enough to hold the PSP and the program image. Since the default
+            // maxSizeInParagraphs already accounts for the PSP and program image, we'll just leave
+            // it at that. Allocation will fail later on when it can't find enough space. That's
+            // okay. We just need to ensure that the size we calculate isn't too small.
+            if (freeSizeInParagraphs >= baseSizeInParagraphs) {
+                size.MaxSizeInParagraphs = freeSizeInParagraphs;
+            }
+        }
+
+        return size;
+    }
+
+    /// <summary>
+    /// Allocates a memory block that's at least as large as the minimum size but may be as large as
+    /// the maximum size if there is a large enough free block available.
+    /// </summary>
+    /// <param name="size">The minimum/maximum size of the block to allocate.</param>
+    /// <returns>
+    /// The allocated <see cref="DosMemoryControlBlock"/>,
+    /// or <c>null</c> if no memory block could be found.
+    /// </returns>
+    private DosMemoryControlBlock? AllocateMemoryRange(AllocRange size) {
+        DosMemoryControlBlock? block = AllocateMemoryBlock(size.MaxSizeInParagraphs);
+        if (block is not null) {
+            return block;
+        }
+
+        DosMemoryControlBlock largestFreeBlock = FindLargestFree();
+        if (largestFreeBlock.IsValid && largestFreeBlock.Size >= size.MinSizeInParagraphs) {
+            block = largestFreeBlock;
+            block.PspSegment = _pspTracker.GetCurrentPspSegment();
```

Comment:
no need to overwrite block. You can modify largestFreeBlock and return it directly.
However, is it OK not to mark it as non free?

Reply:
This function is supposed to allocate it, so I think it's safe to mark it as non-free. It's the same thing that the existing allocation functions do.

I agree that assigning it to `block` isn't technically necessary. The reason I did that is so that I could have just a single return statement with `block` regardless of whether it was `null` or not. Also making the `largestFreeBlock` variable non-free read weird, and I thought it would be slightly confusing.